torch/_refs/__init__.py

```diff
@@ -4501,8 +4527,6 @@ def masked_fill(a: TensorLikeType, mask: TensorLikeType, value: TensorOrNumberLi
             lambda: "Expected `value` to be on same device as `a`",
         )
         value_type = utils.dtype_to_type(value.dtype)
-        if utils.is_cpu_scalar_tensor(value):
-            value = value.item()
```

Comment:
Is this okay? Looks like it doesn't break anything and more tests pass now (due to not having data-dependent control flow). I've also added a repro that was linked from a PR that added this, see `test_masked_fill_device_promotion`.